syntax = "proto3";
package csce662;
import "google/protobuf/empty.proto";

service CoordService {
  rpc Register(ServerRegistration) returns (RegistrationResponse) {}
  rpc Heartbeat(HeartbeatMessage) returns (google.protobuf.Empty) {}
  rpc GetServer(ClientID) returns (ServerInfo) {}
}

service ReplicatorService {
  rpc InformNewSlave(ServerInfo) returns (google.protobuf.Empty) {}
}

// service SynchronizerService {

// }

message ClientID {
  int32 id = 1;
}

enum ServerCapability {
  SNS = 0;
  SNS_REPLICATOR = 1;
  SNS_SYNCHRONIZER = 2;
}

enum ClusterStatus {
  MASTER = 0;
  SLAVE = 1;
}

message RegistrationResponse {
  ClusterStatus status = 1;
}

message HeartbeatMessage {
  int32 cluster_id = 1;
  int32 server_id = 2;
}

message ServerRegistration {
  int32 cluster_id = 1;
  ServerInfo info = 2;
  repeated ServerCapability capabilities = 3;
}

message ServerInfo {
  int32 id = 1;
  string hostname = 2;
  int32 port = 3;
}
